### ISSUE TITLE
GenFacades fails to generate not supported assembly for reference source that contains throw null lambda.

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/arcade#6509]
- [x] This issue is blocking 
- [ ] This issue is causing unreasonable pain

GenFacades not supported assembly fails for reference source files that use the `=> throw null;` syntax instead of a full method body.

```
C:\git\runtime5\artifacts\obj\System.Net.Quic\net6.0-Browser-Debug\System.Net.Quic.notsupported.cs(17,9): error CS8057: Block bodies and expression bodies cannot both be provided. [C:\git\runtime5\src\libraries\System.Net.Quic\src\System.Net.Quic.csproj]

        public QuicListener(IPEndPoint listenEndPoint, System.Net.Security.SslServerAuthenticationOptions sslServerAuthenticationOptions) { throw new System.PlatformNotSupportedException(System.SR.SystemNetQuic_PlatformNotSupported);  }
=> throw null;
```

cc @Anipik @ericstj

**Comment by ericstj** [comment]
We should fix this, but it doesn't seem blocking.

Why do we have `=> throw null;` syntax in the ref source?  I don't think GenAPI produces that.

**Comment by Anipik** [comment]
We can use https://apisof.net/catalog/Microsoft.CodeAnalysis.CSharp.Syntax.MethodDeclarationSyntax.ExpressionBody
to differentiate between the 2 cases and fix the error.

**Comment by ViktorHofer** [comment]
It block developers who want to generate a PNSE assembly from reference sources which contain these lambdas. We have at least one hit in System.Net.Quic.cs. I ticked the "blocking" checkbox as my stance on this is that it blocks a developer who doesn't know anything about PNSE generation in GenFacades.

**Comment by ericstj** [comment]
I understand, I was just pointing out that we have an overriding desire to make our checked in ref-source be a clean diff when running GenAPI so it's probably "the right fix" to make that match what GenAPI emits.  Now I wouldn't be opposed to also having GenAPI emit the terser syntax but that's another feature altogether /cc @safern.

**Comment by ViktorHofer** [comment]
Which style does RefOut emit?

**Comment by ViktorHofer** [comment]
> "the right fix" to make that match what GenAPI emits

The right fix if developers use GenAPI to generate/update the reference source. As we can see by the System.Net.Quic example, developers tend to not follow that workflow.

**Comment by ericstj** [comment]
Sure but I would accept a PR to the repo that made all ref source a clean diff with output of GenAPI.  That tells me we have a preference.

**Comment by safern** [comment]
> Which style does RefOut emit?

I believe `RefOut` doesn't emit any C# code, it just emits an assembly right?

**Comment by ViktorHofer** [comment]
Yeah I missed that the IL produced from an expression or method body is identical.

**Comment by ericstj** [comment]
Correct, once switching to refout we will still use GenAPI to emit the cs file. That can be an input to the PublicApiAnalyzer (and the PNSE source task).

**Comment by ViktorHofer** [comment]
I know this is completely off-topic now but did we ask the roslyn team for a GenAPI equivalent?

**Comment by ericstj** [comment]
@safern is currently investigating. We have discussed with folks from Roslyn in the past and no tool existed but the closest one was the functionality that backed go-to definition.  We indeed want Roslyn APIs to back as much of the GenAPI stack as possible so we get new language features automatically.  We also desire to publicly ship the tool and there is likely a better place for it to live than arcade or runtime.

**Comment by ViktorHofer** [comment]
What would it take to make that a NET6 deliverable for roslyn? IMO this is similar to our ongoing talk with the NuGet team to support restoring duplicate tfm/tpm tuples for net6.

**Comment by ericstj** [comment]
@safern mentioned:
> Seems like the fix might be trivial. Just clearing the expression body here: https://github.com/dotnet/arcade/blob/77daa47917f98e213e0412cbf36ff0ff50475691/src/Microsoft.DotNet.GenFacades/NotSupportedAssemblyGenerator.cs#L127 

> Or we could always clear the method body and always return the method declaration syntax with an expression body instead:https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.methoddeclarationsyntax.withexpressionbody?view=roslyn-dotnet#Microsoft_CodeAnalysis_CSharp_Syntax_MethodDeclarationSyntax_WithExpressionBody_Microsoft_CodeAnalysis_CSharp_Syntax_ArrowExpressionClauseSyntax_

> or just handle each case separately there. 
 
It strikes me as unusual that a block with both a Body and ExpressionBody uses the ExpressionBody, even when the Body was the last value set.  Perhaps that's a Roslyn bug?

**Comment by safern** [comment]
cc: @CyrusNajmabadi @jaredpar for that question

**Comment by jaredpar** [comment]
> It strikes me as unusual that a block with both a Body and ExpressionBody uses the ExpressionBody, even when the Body was the last value set. Perhaps that's a Roslyn bug?

Assume that you are referring to how this case is handled when used in a `Compilation` or `SemanticModel`? If so then only `SyntaxTree` that are produced from our parsers are officially supported in those context. Yes you can, and people often do, produce hand generated `SyntaxTree` and pass them into those contexts. That will often work but if you end up producing a `SyntaxTree` that can't be produced by the parser then the behavior is undefined. In this case the parser can't produce both a `Body` and `ExpressionBody` hence you're in undefined territory here.

**Comment by CyrusNajmabadi** [comment]
>  Perhaps that's a Roslyn bug?

No. Our syntax api makes no policy decisions on your behalf. It does exactly what you ask of it. If you ask to have both a block and expr body, it will happily trust that you know what you are doing. 

Policy decisions happen above this layer. 

You can think of the syntax model as being much closer to a collections api (albeit, strongly typed, and tree oriented), then a high level correctness api.
